hilti/toolchain/src/compiler/jit.cc

```diff
@@ -288,45 +389,37 @@ hilti::Result<std::shared_ptr<const Library>> JIT::_link() {
             HILTI_DEBUG(logging::debug::Driver, util::fmt("saving object file to %s", dbg));
 
             std::error_code ec;
-            hilti::rt::filesystem::copy(_tmpdir->path() / path, dbg,
-                                        hilti::rt::filesystem::copy_options::overwrite_existing,
+            hilti::rt::filesystem::copy(path, dbg, hilti::rt::filesystem::copy_options::overwrite_existing,
                                         ec); // will save into current directory; ignore errors
         }
     }
 
+    // We are using the compiler as a linker here, no need to use a compiler launcher.
+    // Since we are writing to a random temporary file non of this would cache anyway.
     if ( auto rc = _spawnJob(hilti::configuration().cxx, std::move(args)); ! rc )
         return rc.error();
 
     if ( auto rc = _waitForJobs(); ! rc )
         return rc.error();
 
-    // Copy the library to a new location that won't be deleted when JIT has
-    // finished. The library itself will clean up the file when no longer
-    // needed.
-    auto tmp_dir = _make_tmp_directory();
-    if ( ! tmp_dir )
-        return tmp_dir.error();
-
-    auto ext_lib = *tmp_dir / lib.filename();
-
-    try {
-        HILTI_DEBUG(logging::debug::Jit, util::fmt("copying library to %s", ext_lib));
-        hilti::rt::filesystem::create_directory(ext_lib.parent_path());
-        hilti::rt::filesystem::rename(_tmpdir->path() / lib, ext_lib);
-    } catch ( hilti::rt::filesystem::filesystem_error& e ) {
-        return result::Error(e.what());
-    }
+    // Atomically move the temprorary file to its final location. With that
+    // even with concurrent saves to the same final path other processes should
+    // always see a consistent version of the contents of that file.
+    auto lib = hilti::rt::filesystem::temp_directory_path() / util::fmt("__library__%" PRIx64 ".hlto", _hash);
+    std::error_code ec;
+    hilti::rt::filesystem::rename(lib0, lib, ec);
+    if ( ec )
+        rt::fatalError(util::fmt("could not move file %s to final location %s: %s", lib0, lib, ec.message()));
 
     // Instantiate the library object from the file on disk, and set it up
     // to delete the file & its directory on destruction.
-    auto library = std::shared_ptr<const Library>(new Library(ext_lib), [ext_lib](const Library* library) {
-        auto remove = library->remove();
-        if ( ! remove )
-            logger().warning(util::fmt("could not remove JIT library: %s", remove.error()));
-
-        std::error_code ec;
-        if ( ! hilti::rt::filesystem::remove(ext_lib.parent_path(), ec) )
-            logger().warning(util::fmt("could not remove JIT temporary library directory: %s", ec.message()));
+    bool keep_tmps = options().keep_tmps;
+    auto library = std::shared_ptr<const Library>(new Library(lib), [keep_tmps](const Library* library) {
```

Comment:
Could use the new `FileGuard` here as well?

Reply:
There isn't really a scope here to which we can the lifetime of `library` since we pass it a couple layers back up. I'd suggest we leave this as is for this PR (the use of a `shared_ptr` to manage cleanup of `library` also predates this PR).